src/pudl/metadata/classes.py

```diff
@@ -1495,6 +1496,25 @@ def format_df(self, df: pd.DataFrame | None = None, **kwargs: Any) -> pd.DataFra
                 df[key] = PERIODS[period](df[key])
         return df
 
+    def enforce_schema(self, df: pd.DataFrame) -> pd.DataFrame:
+        """Drop columns not in the DB schema and enforce specified types."""
+        expected_cols = pd.Index(self.get_field_names())
+        missing_cols = list(expected_cols.difference(df.columns))
+        if missing_cols:
+            raise ValueError(
+                f"{self.name}: Missing columns found when enforcing table "
+                f"schema: {missing_cols}"
+            )
+        df = self.format_df(df)
+        pk = self.schema.primary_key
+        if pk and not df[df.duplicated(subset=pk)].empty:
+            raise ValueError(
+                f"{self.name} Duplicate primary keys when enforcing schema."
+            )
+        if pk and df.loc[:, pk].isna().any(axis=None):
+            raise ValueError(f"{self.name} Null values found in primary key columns.")
+        return df
+
```

Comment:
Not... strictly related to this PR but I moved this into the `Resource` class so it can be used on any table, not just the FERC 1 tables.  I thought about using it on the EIA-861 tables, but ended up just doing the gentler `apply_pudl_dtypes()` instead. Maybe I should swap this into `_post_process()` instead and see what happens?

Also added the PK NOT NULL check.

Reply:
Uuuughg, I feel like this is a whole major epic / tool-search in the making. But yeah.

I think we need to differentiate between tables with schemas and tables without schemas, and allow either to be loaded into the DB, but specify which we are doing, so that in the case where we have a schema, and things don't match, we get errors.

So yeah, something like this could go in the `SQLiteIOManager`, but maybe it should be happening outside of that in a more general place? It's doing some checks and converting some dtypes as appropriate for Pandas, in such a way that we think the types will also be compatible with SQLite.

The dtypes are so squirrly that this kind of thing has to be done basically every time the df is being passed off to another step where you need to have some certainty about wtf is going on with the types.

Reply:
> The dtypes are so squirrly that this kind of thing has to be done basically every time the df is being passed off to another step where you need to have some certainty about wtf is going on with the types.

Agreed, it seems like pandas to sqlite types are the most uncertain which is why I proposed adding `enforce_schema()` to SQLiteIOManager.

I figure our dtype conversion helper functions will be enough to keep the types straight. If this becomes an issue, we could create a new IO manager that read and writes dataframes as pickle files but dumps the dataframes to sqlite on the way. This would give us more confidence in the dtypes staying the same.

As for this PR, I think we're fine for now. We can continue thinking about the best way to maintain dtypes as we work with the system more.